pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -1019,34 +1038,106 @@ public synchronized void close()
     _logger.info("Closed the metadata manager");
   }
 
-  protected void replaceDocId(ThreadSafeMutableRoaringBitmap validDocIds,
+  /**
+   * Use WLock to make updates on two segments' bitmaps atomically.
+   */
+  protected void replaceDocId(IndexSegment newSegment, ThreadSafeMutableRoaringBitmap validDocIds,
       @Nullable ThreadSafeMutableRoaringBitmap queryableDocIds, IndexSegment oldSegment, int oldDocId, int newDocId,
       RecordInfo recordInfo) {
-    removeDocId(oldSegment, oldDocId);
-    addDocId(validDocIds, queryableDocIds, newDocId, recordInfo);
+    // For SNAPSHOT consistency mode, we can use RLock here. But for simplicity and considering there is only one
```

Comment:
Suggest keeping it consistent. I found it easier to understand if we always use RL for updates and WL for sync

Reply:
kk, separated and commented for more clarity.